File: packages/sdk/src/lib/registry.ts
// registry.ts — HTTP client for the windy-registry service.
//
// Reads the registry base URL from --registry-url, $WINDY_REGISTRY_URL, or
// defaults to https://api.windydrops.com.

import { fetchWithTimeout } from "./http.js";

export interface PublishPayload {
  manifest: Record<string, unknown>;
  bundle_url: string;
  bundle_sha256: string;
}

export interface PublishedDrop {
  drop_id: string;
  version: string;
  manifest: Record<string, unknown>;
  bundle_url: string;
  bundle_sha256: string;
  signature_verified: boolean;
  signer_passport: string | null;
  signer_integrity_band: string | null;
  signer_clearance_level: string | null;
  published_at: string;
}

export class RegistryError extends Error {
  constructor(message: string, readonly status: number, readonly body: unknown) {
    super(message);
    this.name = "RegistryError";
  }
}

/** Look up a drop by id. Returns null on 404; throws RegistryError otherwise. */
export async function fetchDrop(
  opts: { registryUrl: string; dropId: string },
): Promise<Record<string, unknown> | null> {
  const url = `${opts.registryUrl}/api/v1/drops/${encodeURIComponent(opts.dropId)}`;
  const r = await fetchWithTimeout(url, {}, 15_000);
  if (r.status === 404) return null;
  if (!r.ok) {
    throw new RegistryError(`registry GET drop: ${r.status}`, r.status, await r.text());
  }
  return (await r.json()) as Record<string, unknown>;
}

export function resolveRegistryUrl(opts: { registryUrl?: string } = {}): string {
  return (
    opts.registryUrl ||
    process.env.WINDY_REGISTRY_URL ||
    "https://api.windydrops.com"
  ).replace(/\/$/, "");
}

export async function publishToRegistry(opts: {
  registryUrl: string;
  bearerToken: string;
  payload: PublishPayload;
}): Promise<PublishedDrop> {
  const url = `${opts.registryUrl}/api/v1/drops`;
  const r = await fetchWithTimeout(url, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      authorization: `Bearer ${opts.bearerToken}`,
    },
    body: JSON.stringify(opts.payload),
  }, 30_000);
  if (!r.ok) {
    let body: unknown;
    try {
      body = await r.json();
    } catch {
      body = await r.text();
    }
    throw new RegistryError(
      `registry rejected publish: ${r.status}`,
      r.status,
      body,
    );
  }
  return (await r.json()) as PublishedDrop;
}

export async function uploadBundleBytes(opts: {
  registryUrl: string;
  bearerToken: string;
  dropId: string;
  version: string;
  zipBytes: Uint8Array;
}): Promise<string[]> {
  // PUT the bundle zip to the registry, which re-verifies the SHA-256
  // against the published version row and pushes zip + members to R2.
  const url = `${opts.registryUrl}/api/v1/drops/${opts.dropId}/versions/${opts.version}/bundle`;
  const r = await fetchWithTimeout(url, {
    method: "PUT",
    headers: {
      "content-type": "application/zip",
      authorization: `Bearer ${opts.bearerToken}`,
    },
    body: opts.zipBytes,
  }, 120_000);
  if (!r.ok) {
    let body: unknown;
    try {
      body = await r.json();
    } catch {
      body = await r.text();
    }
    throw new RegistryError(
      `registry rejected bundle upload: ${r.status}`,
      r.status,
      body,
    );
  }
  const data = (await r.json()) as { uploaded?: string[] };
  return data.uploaded ?? [];
}

File: python/sdk/src/windy_drops/lib/registry.py
"""registry.py — HTTP client for the windy-registry service."""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Any

import httpx


@dataclass(frozen=True)
class PublishedDrop:
    drop_id: str
    version: str
    manifest: dict[str, Any]
    bundle_url: str
    bundle_sha256: str
    signature_verified: bool
    signer_passport: str | None
    signer_integrity_band: str | None
    signer_clearance_level: str | None
    published_at: str


class RegistryError(RuntimeError):
    def __init__(self, message: str, status: int, body: Any) -> None:
        super().__init__(message)
        self.status = status
        self.body = body


def fetch_drop(*, registry_url: str, drop_id: str) -> dict[str, Any] | None:
    """Look up a drop by id. Returns None on 404; raises RegistryError otherwise."""
    url = f"{registry_url}/api/v1/drops/{drop_id}"
    r = httpx.get(url, timeout=15.0)
    if r.status_code == 404:
        return None
    if r.status_code >= 400:
        raise RegistryError(f"registry GET drop: {r.status_code}", r.status_code, r.text)
    return r.json()


def resolve_registry_url(*, registry_url: str | None = None) -> str:
    return (
        (registry_url or os.environ.get("WINDY_REGISTRY_URL") or "https://api.windydrops.com")
        .rstrip("/")
    )


def upload_bundle_bytes(
    *,
    registry_url: str,
    bearer_token: str,
    drop_id: str,
    version: str,
    zip_bytes: bytes,
) -> list[str]:
    """PUT the bundle zip to the registry, which pushes it to R2.

    Returns the list of uploaded object keys. The registry re-verifies the
    SHA-256 against the published version row, so a tampered zip is rejected.
    """
    url = f"{registry_url}/api/v1/drops/{drop_id}/versions/{version}/bundle"
    r = httpx.put(
        url,
        content=zip_bytes,
        headers={
            "authorization": f"Bearer {bearer_token}",
            "content-type": "application/zip",
        },
        timeout=120.0,
    )
    if r.status_code >= 400:
        try:
            body = r.json()
        except Exception:
            body = r.text
        raise RegistryError(
            f"registry rejected bundle upload: {r.status_code}",
            status=r.status_code,
            body=body,
        )
    return r.json().get("uploaded", [])


def publish_to_registry(
    *,
    registry_url: str,
    bearer_token: str,
    payload: dict[str, Any],
) -> PublishedDrop:
    url = f"{registry_url}/api/v1/drops"
    r = httpx.post(
        url,
        json=payload,
        headers={"authorization": f"Bearer {bearer_token}"},
        timeout=30.0,
    )
    if r.status_code >= 400:
        try:
            body = r.json()
        except Exception:
            body = r.text
        raise RegistryError(
            f"registry rejected publish: {r.status_code}",
            status=r.status_code,
            body=body,
        )
    data = r.json()
    return PublishedDrop(
        drop_id=data["drop_id"],
        version=data["version"],
        manifest=data["manifest"],
        bundle_url=data["bundle_url"],
        bundle_sha256=data["bundle_sha256"],
        signature_verified=data["signature_verified"],
        signer_passport=data.get("signer_passport"),
        signer_integrity_band=data.get("signer_integrity_band"),
        signer_clearance_level=data.get("signer_clearance_level"),
        published_at=data["published_at"],
    )
